core/templates/dev/head/pages/exploration_editor/editor_tab/ResponsesService.js

```diff
@@ -0,0 +1,346 @@
+          // cover most cases.
+          //
+          // TODO(sll): Find a way to make this fully deterministic. This can
+          // probably only occur after we support custom editors for
+          // interactions.
+          var onlyEditsHappened = false;
+          if (oldAnswerChoices.length === newAnswerChoices.length) {
+            onlyEditsHappened = true;
+
+            // Check that no answer choice appears to have been moved.
+            var numAnswerChoices = oldAnswerChoices.length;
+            for (var i = 0; i < numAnswerChoices; i++) {
+              for (var j = 0; j < numAnswerChoices; j++) {
+                if (i !== j &&
+                    oldAnswerChoices[i].val === newAnswerChoices[j].val) {
+                  onlyEditsHappened = false;
+                  break;
+                }
+              }
+            }
+          }
+
+          var oldChoiceStrings = oldAnswerChoices.map(function(choice) {
+            return choice.val;
+          });
+          var newChoiceStrings = newAnswerChoices.map(function(choice) {
+            return choice.val;
+          });
+
+          _answerGroups.forEach(function(answerGroup, answerGroupIndex) {
+            var newRules = angular.copy(answerGroup.rules);
+            newRules.forEach(function(rule) {
+              for (var key in rule.inputs) {
+                var newInputValue = [];
+                rule.inputs[key].forEach(function(item) {
+                  var newIndex = newChoiceStrings.indexOf(item);
+                  if (newIndex !== -1) {
+                    newInputValue.push(item);
+                  } else if (onlyEditsHappened) {
+                    var oldIndex = oldChoiceStrings.indexOf(item);
+                    if (oldIndex !== -1) {
+                      newInputValue.push(newAnswerChoices[oldIndex].val);
+                    }
+                  }
+                });
+                rule.inputs[key] = newInputValue;
+              }
+            });
+
+            _updateAnswerGroup(answerGroupIndex, {
+              rules: newRules
+            });
+          });
+        }
+      },
+      getAnswerGroups: function() {
+        return angular.copy(_answerGroups);
+      },
+      getAnswerGroup: function(index) {
+        return angular.copy(_answerGroups[index]);
+      },
+      getAnswerGroupCount: function() {
+        return _answerGroups.length;
+      },
+      getDefaultOutcome: function() {
+        return angular.copy(_defaultOutcome);
+      },
+      getConfirmedUnclassifiedAnswers: function() {
+        return angular.copy(_confirmedUnclassifiedAnswers);
+      },
+      // This registers the change to the handlers in the list of changes, and
+      // also updates the states object in explorationStatesService.
+      save: function(newAnswerGroups, defaultOutcome) {
+        _saveAnswerGroups(newAnswerGroups);
+        _saveDefaultOutcome(defaultOutcome);
+      }
+    };
+  }
+]);
```

Comment:
Hey @shubha1593  i have a newline character after this line but it is not showing up here 
cannot commit that as git says working directory clean

Reply:
Hey @GaganSuneja, yup the web editor doesn't show the newline(empty line) at the end of the file. But it's present, I checked, so no worries :)

Reply:
@GaganSuneja, also just so you know, the way I checked it was to pull in your PR and open the file in my local editor, where the newline was present. 
![newline_eof](https://user-images.githubusercontent.com/7807557/31092123-0454b4fa-a7cb-11e7-8842-7dc0765af9e4.jpeg)

Also if you're using sublime, then you can ensure that a newline is present at the end of file when you save by having these two lines in your user settings:
"default_line_ending": "unix",
"ensure_newline_at_eof_on_save": true

And if github doesn't show the blank line after the end of file, and you don't get any kind of warnings sign, then most likely, it's fine. Your "files changed" should somewhat look like [this](https://github.com/oppia/oppia/pull/3925/files), then it's most likely okay :)